server/src/main/java/com/cloud/server/ManagementServerImpl.java

```diff
@@ -3938,7 +3938,12 @@ public void doInTransactionWithoutResult(final TransactionStatus status) {
                     }
                     // update password for this host
                     final DetailVO nv = _detailsDao.findDetail(h.getId(), ApiConstants.USERNAME);
-                    if (nv.getValue().equals(command.getUsername())) {
+                    if (nv == null) {
+                        final DetailVO nvu = new DetailVO(h.getId(), ApiConstants.USERNAME, command.getUsername());
```

Comment:
username from the command object can be null or empty ?

Reply:
@sureshanaparti username and password are mandatory parameters.. They cant be null

```
(local) 🐱 > update hostpassword hostid=50c8-b73a-4eaf-bbab-7a70
💩 Missing required parameters:  password, username
```

Reply:
@ravening empty username string ?

Reply:
> @ravening empty "" (or) "   " username string ? 

@sureshanaparti those are non null values

Reply:
@sureshanaparti  Even if I dont pass username, I dont get NPE

```
 update hostpassword hostid=50436ac8-b73a-4eaf-bbab-7adbd9a50870 password=cloudstack username=
🙈 Error: (HTTP 431, error code 4350) The username is not same for the hosts..
```

Reply:
@ravening username is not null, but empty string, so can you add check for that ? Ex: "_StringUtils.isBlank(command.getUsername())_"

Reply:
> @ravening username is not null, but empty string, so can you add check for that ? Ex: "_StringUtils.isBlank(command.getUsername())_" 

@sureshanaparti but why? For some hosts, username and password can be empty string

Reply:
@ravening if you pass username, then it shouldn't be empty, password can be empty. can you confirm if username  is persisted with empty value in the _host_details_ table?

Reply:
@sureshanaparti 

```
(local) 🐱 > update hostpassword hostid=50436ac8-b73a-4eaf-bbab-7adbd9a50870 password=cloudstack username="  "
🙈 Error: (HTTP 431, error code 4350) The username is not same for the hosts..
(local) 🐱 >
(local) 🐱 > update hostpassword hostid=50436ac8-b73a-4eaf-bbab-7adbd9a50870 password=cloudstack username=""
🙈 Error: (HTTP 431, error code 4350) The username is not same for the hosts..
(local) 🐱 >
```

Reply:
@ravening can you test your changes when "_nv == null_" with above username?

Reply:
@sureshanaparti 
```
(local) 🐱 > update hostpassword hostid=f16d5515-8c53-4873-b395-dc439f5c67d6 password=cloudstack username=" "
{
  "success": true
}
```

this is when nv is null

Reply:
can you share the entries in _host_details_ table of that test?

Reply:
> can you share the entries in _host_details_ table of that test?

I think it's not going anywhere further. Can you let me know what exactly you are looking for?

Reply:
> can you share the entries in _host_details_ table of that test?

@sureshanaparti  the last result I pasted was tested when host_details entry were removed for that host

Reply:
@ravening No point in saving password with username empty. Hope you got it. Wanted to confirm the same based on the entries persisted in _host_details_ table for your test. Can you share these? Thanks.

Reply:
@sureshanaparti @ravening we had the issue in the past that Removed kvm host is added back to cloudstack when cloudstack-agent is started. in that case, there is no username/password saved in host_details for the host (because the entries are removed from db when host is removed, but not added back).